src/smoke_tests/threads.c

```diff
@@ -0,0 +1,124 @@
+#include <stdlib.h>
+#include <unistd.h>
+#include <stdio.h>
+#include <pthread.h>
+#include <papi.h>
+#define NUM_PTHREADS 2
+#define NUM_EVENTS 2 
+
+void *
+Thread(void *arg)
+{
+    int retval, i;
+    long long       values[NUM_EVENTS];
+    int             EventSet = PAPI_NULL;
+    int             events[NUM_EVENTS];
+    char            *EventName[] = { "PAPI_TOT_CYC", "PAPI_TOT_INS" };
+    
+    int thread;
+    thread = *(int *) arg;
+ 
+    retval = PAPI_register_thread(  );
+    if ( retval != PAPI_OK ) {
+        printf("ERROR: PAPI_register_thread: %d: %s\n", retval, PAPI_strerror(retval));
+        exit(EXIT_FAILURE);
+    }
+
+    for( i = 0; i < NUM_EVENTS; i++ ) {
+        retval = PAPI_event_name_to_code( EventName[i], &events[i] );
+        if ( retval != PAPI_OK ) {
+            printf("ERROR: PAPI_event_name_to_code: %d: %s\n", retval, PAPI_strerror(retval));
+            exit(EXIT_FAILURE);
+        }
+    }
+
+    retval = PAPI_create_eventset( &EventSet );
+    if ( retval != PAPI_OK ) {
+        printf("ERROR: PAPI_create_eventset: %d: %s\n", retval, PAPI_strerror(retval));
+        exit(EXIT_FAILURE);
+    }
+
+    retval = PAPI_add_events( EventSet, events, NUM_EVENTS );
+    if ( retval != PAPI_OK ) {
+        printf("ERROR: PAPI_add_events: %d: %s\n", retval, PAPI_strerror(retval));
+        exit(EXIT_FAILURE);
+    }
+
+    retval = PAPI_start( EventSet );
+    if ( retval != PAPI_OK ) {
+        printf("ERROR: PAPI_start: %d: %s\n", retval, PAPI_strerror(retval));
+        exit(EXIT_FAILURE);
+    }        
+
+    // do work
+    sleep(3);
+                
+    retval = PAPI_stop( EventSet, values );
+    if ( retval != PAPI_OK ) {
+        printf("ERROR: PAPI_stop: %d: %s\n", retval, PAPI_strerror(retval));
+        exit(EXIT_FAILURE);
+    }
+                
+    for( i = 0; i < NUM_EVENTS; i++ ) {           
+        printf( "%12lld \t\t --> %s  (thread %d) \n", values[i],
+                EventName[i], thread );
+    }
+
+    retval = PAPI_unregister_thread(  );
+    if ( retval != PAPI_OK ) {
+        printf("ERROR: PAPI_unregister_thread: %d: %s\n", retval, PAPI_strerror(retval));
+        exit(EXIT_FAILURE);
+    }
+
```

Comment:
Should destroy the eventset too.

Reply:
AFAIK, PAPI_shutdown(), which I included, cleans up and frees event sets.